python/python_common/python_utils.cpp

```diff
@@ -0,0 +1,236 @@
+              -rank,
+              ", but got: ",
+              order);
+        } else {
+          NVF_CHECK(
+              order < rank,
+              "defineTensor stride_order argument is out of range, expects < ",
+              rank,
+              ", but got: ",
+              order);
+        }
+      });
+
+  // Then, normalize negative values.
+  std::unordered_set<int64_t> order_set;
+  order_set.reserve(rank);
+  std::transform(
+      stride_order.begin(),
+      stride_order.end(),
+      std::inserter(order_set, order_set.begin()),
+      [rank](int64_t order) { return wrapDim(order, rank); });
+
+  NVF_CHECK(
+      order_set.size() == stride_order.size(),
+      "defineTensor got duplicated stride_order entries: " +
+          toDelimitedString(stride_order));
+}
+
+std::vector<bool> getExpanded(
+    const std::vector<int64_t>& shape,
+    const std::vector<std::optional<bool>>& contiguity,
+    const std::vector<int64_t>& stride_order) {
+  size_t rank = shape.size();
+  std::vector<bool> is_expand(rank);
+  for (size_t index : arange(rank)) {
+    // since contiguity vector is given to the corresponding order in alloc
+    // domain, while is_expand is given to root domain, we need to map it
+    // correctly with `contig_index` and `index`.
+    //
+    // stride_order[i] indicates that:
+    //   `logical_domain[i]` maps to `alloc_domain[rank - 1 - stride_order_[i]]`
+    //
+    // Hence `index` on root domain would be corresponding to the contiguity
+    // index `contig_index = rank - 1 - stride_order[index]`
+    const size_t contig_index = stride_order.empty()
+        ? index
+        : rank - 1 - static_cast<size_t>(stride_order[index]);
+    const bool is_broadcast = !contiguity[contig_index].has_value();
+    const bool has_non_broadcast_size = (shape[index] != 1);
+    // A root dimension is expand dimension if:
+    //   The dimension is marked a broadcast; and
+    //   The dimension has an expanded extent.
+    is_expand[index] = is_broadcast && has_non_broadcast_size;
+  }
+  return is_expand;
+}
+
+std::vector<std::optional<bool>> getContiguityVec(
+    const std::vector<int64_t>& shape,
+    const std::vector<int64_t>& stride_order,
+    bool contiguity) {
+  const auto rank = static_cast<int64_t>(shape.size());
+  std::vector<std::optional<bool>> contiguity_vec(rank);
+  for (const auto index : arange(rank)) {
+    const auto contig_index =
+        stride_order.empty() ? index : rank - 1 - stride_order[index];
+    if (shape[index] == 1) {
+      contiguity_vec[contig_index] = std::nullopt;
+    } else {
+      contiguity_vec[contig_index] = contiguity;
+    }
+  }
+  return contiguity_vec;
+}
+
+std::vector<int64_t> getTensorViewBuilderSizes(
+    const std::vector<int64_t>& sizes,
+    bool static_sizes) {
+  // TensorViewBuilder assumes any dim with a compile time constant
+  // size == 1 is a broadcast axis, symbolic sizes are
+  // identified by -1, and size == 0 is not supported.
```

Comment:
nitpick: the comment here mentioned about `size == 0` is not supported, yet we do allow 0 in `sizes` in the logic below.